client/services/ReinforcementGeneratorServiceSpec.js

```diff
@@ -0,0 +1,373 @@
+      inputFunctionName: null,
+      outputFunctionName: null,
+      mainFunctionName: 'mockMainFunction',
+      correctnessTests: [{
+        input: 'task_1_correctness_test_1',
+        allowedOutputs: [true],
+        tag: 'test1'
+      }, {
+        input: 'task_1_correctness_test_2',
+        allowedOutputs: [true],
+        tag: 'test2'
+      }],
+      buggyOutputTests: [],
+      performanceTests: []
+    }];
+
+    var taskDict2 = [{
+      instructions: [''],
+      prerequisiteSkills: [''],
+      acquiredSkills: [''],
+      inputFunctionName: null,
+      outputFunctionName: null,
+      mainFunctionName: 'mockMainFunction',
+      correctnessTests: [{
+        input: 'task_2_correctness_test_1',
+        allowedOutputs: [true],
+        tag: 'test1'
+      }, {
+        input: 'task_2_correctness_test_2',
+        allowedOutputs: [true],
+        tag: 'test2'
+      }, {
+        input: 'task_2_correctness_test_3',
+        allowedOutputs: [true],
+        tag: 'test3'
+      }],
+      buggyOutputTests: [],
+      performanceTests: []
+    }];
+
+    var taskDictMultCases = [{
+      instructions: [''],
+      prerequisiteSkills: [''],
+      acquiredSkills: [''],
+      inputFunctionName: null,
+      outputFunctionName: null,
+      mainFunctionName: 'mockMainFunction',
+      correctnessTests: [{
+        input: 'task_3_correctness_test_1',
+        allowedOutputs: [true],
+        tag: 'test1'
+      }, {
+        input: 'task_3_correctness_test_2',
+        allowedOutputs: [true],
+        tag: 'test1'
+      }, {
+        input: 'task_3_correctness_test_3',
+        allowedOutputs: [true],
+        tag: 'test1'
+      }],
+      buggyOutputTests: [],
+      performanceTests: []
+    }];
+
+    testTask = taskDict.map(function(task) {
+      return TaskObjectFactory.create(task);
+    });
+
+    testTask2 = taskDict2.map(function(task) {
+      return TaskObjectFactory.create(task);
+    });
+
+    testTask3 = taskDictMultCases.map(function(task) {
+      return TaskObjectFactory.create(task);
+    });
+  }));
+
+  describe('getReinforcement', function() {
+    it('copies and updates previous reinforcement if from the same task',
+      function() {
+        // Simulate most recent snapshot
+        var codeEvalResult = CodeEvalResultObjectFactory.create(
+          'some code', 'some output', [[false, false]],
+          [], [], null, null);
+        var feedback = FeedbackGeneratorService.getFeedback(
+          testTask, codeEvalResult, [0, 1, 2, 3, 4]);
+
+        TranscriptService.recordSnapshot(null, codeEvalResult, feedback);
+
+        var pastFailedCases = feedback.getReinforcement().getPastFailedCases();
+        var passedTags = feedback.getReinforcement().getPassedTags();
+        var test1 = testTask[0].getCorrectnessTests()[0].getStringifiedInput();
+        var test2 = testTask[0].getCorrectnessTests()[1].getStringifiedInput();
+
+        expect(pastFailedCases[test1]).toEqual(false);
```

Comment:
Would it make sense to just compare pastFailedCases directly to its expected value? Ditto elsewhere.